src/Bicep.LangServer/Configuration/BicepConfigChangeHandler.cs

```diff
@@ -0,0 +1,84 @@
+using Newtonsoft.Json.Linq;
+using Newtonsoft.Json.Schema;
+using OmniSharp.Extensions.LanguageServer.Protocol;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using static Bicep.Core.Diagnostics.DiagnosticBuilder;
+
+namespace Bicep.LanguageServer.Configuration
+{
+    public class BicepConfigChangeHandler : IBicepConfigChangeHandler
+    {
+        private JSchema bicepConfigSchema;
+        private readonly IFileResolver fileResolver;
+
+        public BicepConfigChangeHandler(IFileResolver fileResolver)
+        {
+            this.fileResolver = fileResolver;
+            var assembly = Assembly.GetExecutingAssembly();
+            string manifestResourceName = assembly.GetManifestResourceNames().Where(p => p.EndsWith("bicepconfig.schema.json", StringComparison.Ordinal)).First();
+            Stream? stream = assembly.GetManifestResourceStream(manifestResourceName);
+            var streamReader = new StreamReader(stream ?? throw new ArgumentNullException("Stream is null"), Encoding.Default);
+
+            bicepConfigSchema = JSchema.Parse(streamReader.ReadToEnd());
+        }
+
+        public void RetriggerCompilationOfSourceFilesInWorkspace(ICompilationManager compilationManager, FileEvent bicepConfigFileEvent, IWorkspace workspace)
+        {
+            Uri bicepConfigUri = bicepConfigFileEvent.Uri.ToUri();
+
+            if (ShouldUpsertCompilation(bicepConfigFileEvent.Type, bicepConfigUri))
+            {
+                foreach (ISourceFile sourceFile in workspace.GetSourceFilesForDirectory(bicepConfigUri))
+                {
+                    Uri uri = sourceFile.FileUri;
+
+                    if (fileResolver.TryRead(uri, out string? bicepFileContents, out ErrorBuilderDelegate _) &&
+                        !string.IsNullOrWhiteSpace(bicepFileContents))
+                    {
+                        compilationManager.UpsertCompilation(DocumentUri.From(uri), null, bicepFileContents);
+                    }
+                }
+            }
+        }
+
+        private bool ShouldUpsertCompilation(FileChangeType fileChangeType, Uri bicepConfigUri)
+        {
+            // When fileChangeType is "Changed", we don't want to retrigger compilation for every key stroke.
+            // We'll check the validity of file against the schema and retrigger compilation only if the file is valid
+            if (fileChangeType == FileChangeType.Changed)
+            {
+                return fileResolver.TryRead(bicepConfigUri, out string? bicepConfigFileContents, out ErrorBuilderDelegate _) &&
+                    bicepConfigFileContents is not null &&
+                    IsBicepConfigValid(bicepConfigFileContents);
+            }
+
+            return true;
+        }
+
+        private bool IsBicepConfigValid(string bicepConfig)
+        {
+            try
+            {
+                JObject jObject = JObject.Parse(bicepConfig);
+                return jObject.IsValid(bicepConfigSchema);
```

Comment:
Do we need to do server-side schema validation? Performance on schema validation can be quite slow.

Reply:
I did this to avoid re triggering compilation on every key stroke. I couldn't think of a better way to handle the scenario. I don't quite understand your question on moving to bicep core libraries. Do you mean the validation logic?

Reply:
So at present, `ConfigHelper` handles the loading and reading of config deep inside the semantic model:

https://github.com/Azure/bicep/blob/3e6e5813dc0962984980f168d94b4d6ac2d3d703/src/Bicep.Core/Semantics/SemanticModel.cs#L68
->
https://github.com/Azure/bicep/blob/3e6e5813dc0962984980f168d94b4d6ac2d3d703/src/Bicep.Core/Analyzers/Linter/LinterAnalyzer.cs#L26
-> 
https://github.com/Azure/bicep/blob/3e6e5813dc0962984980f168d94b4d6ac2d3d703/src/Bicep.Core/Configuration/ConfigHelper.cs#L26
->
https://github.com/Azure/bicep/blob/3e6e5813dc0962984980f168d94b4d6ac2d3d703/src/Bicep.Core/Configuration/ConfigHelper.cs#L51-L57

I'm wondering if the creator of the `SemanticModel` instance should be responsible for creating the config and passing it in as a constructor argument, to avoid unneccessary file reads. The `ConfigHelper` itself could also be responsible for validating the config file, rather than relying on a JSON schema validator.

Reply:
Updated as part of this commit:
https://github.com/Azure/bicep/commit/7c53107f3fb09a3e6403cf76b8ee6a876c13806c